### PR TITLE
Add a page that runs PyTorch's memory snapshot viewer to pytorch.org

### PR DESCRIPTION
**Pull request by zdevito**
[PyTorch Memory Snapshots](https://zdevito.github.io/2022/12/09/memory-traces.html) are a way to see how CUDA memory is used over time in a PyTorch process. We have an interactive viewer inside the pytorch repo that allows users to drag/drop a pickled snapshot file onto the viewer, which is a single page javascript application that works entirely locally. This PR hosts a copy of that app on pytorch.org so that uses can go to pytorch.org/memory_viz to use the viewer. 

Future PR's will update the documentation to describe how to use the viewer. 

MemoryViz.js is a chunk of javascript in the PyTorch repo which depends on d3. Including it here will increase the surface of javascript on pytorch.org for cross site scripting. Currently we do not have any authentication hosted on pytorch.org so this increase in javascript surface area does not seem like it would be an issue, but should be revisited if we add some authenticated javascript to pytorch.org
Screenshot 2023-08-02 at 5 33 28 PM
Screenshot 2023-08-02 at 5 33 45 PM
Screenshot 2023-08-02 at 5 33 19 PM

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
Sounds good!
Should this be linked from the main page one way or the other?
Also should we have a stable and main version of this?

**Comment by zdevito** [comment]
> Should this be linked from the main page one way or the other?

Maybe we should link it in the docs describing how to view it, and add it to the API documentation for the functions related to saving snapshots.

> Also should we have a stable and main version of this?

We could point the main page version to the current stable release version of MemoryViz.js with another one pointing to the nightly. Since there hasn't been a release yet with MemoryViz.js, the only option now is to point it to the version on main.

File: memory_viz.html [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<!DOCTYPE html>
+<html>
+<body>
+<script type="module" src="https://cdn.jsdelivr.net/gh/pytorch/pytorch@main/torch/utils/viz/MemoryViz.js"></script>
```
Comment (malfet) [inline memory_viz.html:4]: Why do you need different CDN than main github?
```suggestion
<script type="module" src="https://raw.githubusercontent.com/pytorch/pytorch/main/torch/utils/viz/MemoryViz.js"></script>
```

Comment (zdevito) [reply memory_viz.html:4]: When served from the `raw` endpoint it has the wrong MIME type: `TypeError: 'text/plain' is not a valid JavaScript MIME type.`. This is why things like `jsdelivr` were  created which just serve text files from GitHub as javascript files.

**Comment by albanD** [comment]
I agree that we should update the pytorch.org/docs page to point to this and we can do the move to have both a stable and main after the release.
I think we want the docs page to point to the stable version in general so it would be best to name this unique version stable so that we can link it from the doc. We can add the new main version later on.

**Comment by albanD** [comment]
I agree that we should update the pytorch.org/docs page to point to this and we can do the move to have both a stable and main after the release.
I think we want the docs page to point to the stable version in general so it would be best to name this unique version stable so that we can link it from the doc. We can add the new main version later on.

**Review by albanD — APPROVED** [review]
SGTM!

Once things are released, we most likely want to move this to point to the stable version and we can have a separate page here to point to latest main.

**Comment by zdevito** [comment]
In discussion with @albanD , we will merge the simple reference to pytorch/main for now so that docs for the next PyTorch release have somewhere to point to get to the viewer. Once we want to have a stable vs main version of the viewer we will figure out a mechanism for hosting both.

**Pull request merged**